File: sdk/python/src/aether/graph.py
from __future__ import annotations

from dataclasses import dataclass

from .ingest import extract_entities
from .text import cosine, embed, tokenize
from .types import IndexedChunk


def _norm_entity(e: str) -> str:
    return re_sub(e)


def re_sub(e: str) -> str:
    import re

    return re.sub(r"[^a-z0-9]+", " ", e.lower()).strip()


@dataclass
class SemanticGraph:
    entity_to_chunks: dict[str, set[str]]
    chunk_to_entities: dict[str, list[str]]


def build_graph(chunks: list[IndexedChunk]) -> SemanticGraph:
    entity_to_chunks: dict[str, set[str]] = {}
    chunk_to_entities: dict[str, list[str]] = {}
    for c in chunks:
        ents = list(dict.fromkeys(_norm_entity(e) for e in c.entities if _norm_entity(e)))
        chunk_to_entities[c.id] = ents
        for e in ents:
            entity_to_chunks.setdefault(e, set()).add(c.id)
    return SemanticGraph(entity_to_chunks, chunk_to_entities)


def seed_entities(query: str, graph: SemanticGraph) -> list[str]:
    from_lex = [_norm_entity(e) for e in extract_entities(query)]
    tokens = tokenize(query)
    seeds: list[str] = []
    seen: set[str] = set()
    for e in from_lex:
        if e in graph.entity_to_chunks and e not in seen:
            seen.add(e)
            seeds.append(e)
    for ent in graph.entity_to_chunks:
        if len(ent) < 3:
            continue
        if any(t in ent or ent in t for t in tokens) and ent not in seen:
            seen.add(ent)
            seeds.append(ent)
        if len(seeds) >= 12:
            break
    return seeds[:12]


def expand_graph(
    query: str,
    seeds: list[str],
    graph: SemanticGraph,
    chunks_by_id: dict[str, IndexedChunk],
    hops: int = 2,
) -> tuple[list[str], list[str], int]:
    q_emb = embed(query)
    visited_chunks: set[str] = set()
    visited_entities: set[str] = set(seeds)
    frontier = list(seeds)
    for _ in range(hops):
        nxt: list[str] = []
        for ent in frontier:
            chunk_ids = graph.entity_to_chunks.get(ent)
            if not chunk_ids:
                continue
            ranked = []
            for cid in chunk_ids:
                ch = chunks_by_id.get(cid)
                if ch:
                    ranked.append((cid, cosine(q_emb, ch.embedding)))
            ranked.sort(key=lambda x: x[1], reverse=True)
            for cid, _score in ranked[:6]:
                if cid in visited_chunks:
                    continue
                visited_chunks.add(cid)
                for e in graph.chunk_to_entities.get(cid, []):
                    if e not in visited_entities:
                        visited_entities.add(e)
                        nxt.append(e)
        frontier = nxt[:16]
        if not frontier:
            break
    return list(visited_chunks), list(visited_entities), hops

File: src/lib/rag/graph.ts
import type { IndexedChunk } from "./types";
import { cosine, embed, tokenize } from "./text";
import { extractEntities } from "./ingest";

/**
 * Relation-free semantic graph, inspired by LinearRAG (ICLR'26) concepts:
 * entity ↔ chunk bipartite links, no LLM-extracted relations, linear
 * construction. Original implementation — not a copy of the research repo.
 */
export interface SemanticGraph {
  entityToChunks: Map<string, Set<string>>;
  chunkToEntities: Map<string, string[]>;
  entityNorm: Map<string, string>;
}

function normEntity(e: string): string {
  return e.toLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
}

export function buildGraph(chunks: IndexedChunk[]): SemanticGraph {
  const entityToChunks = new Map<string, Set<string>>();
  const chunkToEntities = new Map<string, string[]>();
  const entityNorm = new Map<string, string>();
  for (const c of chunks) {
    const ents = c.entities.map(normEntity).filter(Boolean);
    const unique = [...new Set(ents)];
    chunkToEntities.set(c.id, unique);
    for (const e of unique) {
      entityNorm.set(e, e);
      let set = entityToChunks.get(e);
      if (!set) {
        set = new Set();
        entityToChunks.set(e, set);
      }
      set.add(c.id);
    }
  }
  return { entityToChunks, chunkToEntities, entityNorm };
}

export function seedEntities(query: string, graph: SemanticGraph): string[] {
  const fromLex = extractEntities(query).map(normEntity);
  const tokens = tokenize(query);
  const seeds = new Set<string>();
  for (const e of fromLex) {
    if (graph.entityToChunks.has(e)) seeds.add(e);
  }
  for (const [ent] of graph.entityToChunks) {
    if (ent.length < 3) continue;
    if (tokens.some((t) => ent.includes(t) || t.includes(ent))) seeds.add(ent);
  }
  return [...seeds].slice(0, 12);
}

export function expandGraph(opts: {
  query: string;
  seeds: string[];
  graph: SemanticGraph;
  chunksById: Map<string, IndexedChunk>;
  hops?: number;
}): { chunkIds: string[]; expandedEntities: string[]; hops: number } {
  const hops = opts.hops ?? 2;
  const qEmb = embed(opts.query);
  const visitedChunks = new Set<string>();
  const visitedEntities = new Set<string>(opts.seeds);
  let frontier = [...opts.seeds];

  for (let h = 0; h < hops; h++) {
    const next: string[] = [];
    for (const ent of frontier) {
      const chunkIds = opts.graph.entityToChunks.get(ent);
      if (!chunkIds) continue;
      const ranked = [...chunkIds]
        .map((id) => {
          const ch = opts.chunksById.get(id);
          if (!ch) return null;
          return { id, score: cosine(qEmb, ch.embedding) };
        })
        .filter((x): x is { id: string; score: number } => x !== null)
        .sort((a, b) => b.score - a.score)
        .slice(0, 6);
      for (const r of ranked) {
        if (visitedChunks.has(r.id)) continue;
        visitedChunks.add(r.id);
        const ents = opts.graph.chunkToEntities.get(r.id) ?? [];
        for (const e of ents) {
          if (!visitedEntities.has(e)) {
            visitedEntities.add(e);
            next.push(e);
          }
        }
      }
    }
    frontier = next.slice(0, 16);
    if (frontier.length === 0) break;
  }

  return {
    chunkIds: [...visitedChunks],
    expandedEntities: [...visitedEntities],
    hops,
  };
}
